system/lib/pthread/pthread_create.c

```diff
@@ -116,8 +120,10 @@ int __pthread_create(pthread_t* restrict res,
   new->locale = &libc.global_locale;
 
   // Allocate memory for thread-local storage and initialize it to zero.
-  new->tsd = malloc(PTHREAD_KEYS_MAX * sizeof(void*));
-  memset(new->tsd, 0, PTHREAD_KEYS_MAX * sizeof(void*));
+  if (__pthread_tsd_size) {
```

Comment:
if pthread_tsd_size is 0, that's when the TLS area will have been allocated with the stack area? is it expected that it will eventually stay as 0?

Reply:
__pthread_tsd_size is the size of area for dynamic pthread_key_create() slots.  Nothing to do with C/C++ static TLS data.

Reply:
ah ok, I guess I was thinking of #16948.
So if this is 0, the slots will be allocated on-demand when pthread_key_create() is called?

Reply:
If its zero it means that the weak version of `__pthread_tsd_size` was used and `pthread_key_create` was not included at link time (pthread_key_create.c contains the strong/non-zero version of `__pthread_tsd_size`):

https://github.com/emscripten-core/emscripten/blob/592f67ff77fea404f07f901b37ec4ff5eb017a25/system/lib/libc/musl/src/thread/pthread_key_create.c#L7

Reply:
got it, thanks.